### ISSUE TITLE
Concurrent docker volume rm/ls commands prevent creation of new volume with same name

### ISSUE THREAD

**Issue by darnells** [docker/docker#21044]
If `volume ls` is called while a volume is actively being removed, the `volume ls` output is delayed until the `volume rm` command completes and a new volume with the same name cannot be created until the daemon is restarted.

Output of `docker version`:

```
Client:
 Version:      1.10.2
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   c3959b1
 Built:        Mon Feb 22 21:37:01 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.10.2
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   c3959b1
 Built:        Mon Feb 22 21:37:01 2016
 OS/Arch:      linux/amd64
```

Output of `docker info`:

```
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 2
Server Version: 1.10.2
Storage Driver: aufs
 Root Dir: /media/ext4/.docker/aufs
 Backing Filesystem: extfs
 Dirs: 69
 Dirperm1 Supported: true
Execution Driver: native-0.2
Logging Driver: json-file
Plugins: 
 Volume: local
 Network: null host bridge
Kernel Version: 3.19.0-32-generic
Operating System: Ubuntu 14.04.4 LTS
OSType: linux
Architecture: x86_64
CPUs: 8
Total Memory: 31.39 GiB
Name: my-host
ID: EIGG:TGSQ:I7TE:MSRZ:AM3Y:67EF:L4H2:CZ2M:IZ4Z:PB7D:LX3T:2YQB
Debug mode (server): true
 File Descriptors: 10
 Goroutines: 31
 System Time: 2016-03-08T16:11:40.974030014-06:00
 EventsListeners: 0
 Init SHA1: 3e247d0d32543488f6e70fbb7c806203f3841d1b
 Init Path: /usr/lib/docker/dockerinit
 Docker Root Dir: /media/ext4/.docker
WARNING: No swap limit support
```

Provide additional environment details (AWS, VirtualBox, physical, etc.):
Physical computer. DOCKER_OPTS in `/etc/default/docker` redirects docker to larger ext4-formatted disk with option `-g`.

List the steps to reproduce the issue:
1. In Shell 1, run `docker volume rm my-volume`. In my case, my-volume contains 1.2 million files totaling 60 GB.
2. In Shell 2 while `volume rm` is running, run `docker volume ls`. When `volume rm` is finished, `volume ls` returns and lists my-volume in stdout.
3. Repeat `docker volume ls`. my-volume is no longer listed.
4. Run `docker volume create --name my-volume`. The exit code for `volume create` is 0 and prints my-volume to stdout.
5. Run `docker volume ls`. my-volume is not listed.
6. Restart the daemon and run `docker volume create --name my-volume` and `docker volume ls`. my-volume is now listed and can be mounted to a new container.

Describe the results you received:
See description in "List the steps to reproduce the issue".

Describe the results you expected:
I expect that Steps 1-5 should successfully create a volume without requiring a daemon restart.

Provide additional info you think is important:
Debug output from daemon for Steps 1-5. No indication of error is reported.

```
DEBU[0884] Calling DELETE /v1.22/volumes/my-volume 
DEBU[0884] DELETE /v1.22/volumes/my-volume       
DEBU[0884] Getting volume reference for name: my-volume 
DEBU[0884] Removing volume reference: driver local, name my-volume 
DEBU[0894] Calling GET /v1.22/volumes                   
DEBU[0894] GET /v1.22/volumes                           
DEBU[1162] Calling GET /v1.22/version                   
DEBU[1162] GET /v1.22/version                           
DEBU[1184] Calling GET /v1.22/info                      
DEBU[1184] GET /v1.22/info                              
DEBU[1822] Calling GET /v1.22/volumes                   
DEBU[1822] GET /v1.22/volumes                           
DEBU[1881] Calling POST /v1.22/volumes/create           
DEBU[1881] POST /v1.22/volumes/create                   
DEBU[1881] form data: {"Driver":"local","DriverOpts":{},"Name":"my-volume"} 
DEBU[2053] Calling GET /v1.22/volumes                   
DEBU[2053] GET /v1.22/volumes
```

**Comment by cpuguy83** [comment]
Thanks for reporting.
Looking at removing the local volume cache which should alleviate this.
I'll have a PR soon.

**Comment by cpuguy83** [comment]
Closing, as I believe this is resolved by #21432

Please ping here if it's still a problem on master.
